### PR TITLE
refactor ssm command waiting and prompt for instances

### PR DESCRIPTION
**Pull request by burnsed**
It appears the the reason ssm commands are failing is because sometimes the command invocation is not available immediately after running `send_command`. Additionally, `run_with_timeout` did not appear to respect the timeout option passed in.

This change refactors `run_with_timeout` to take into account the timeout and handle cases when command invocations are not yet available.

This change also pulls out the instance selection from `prompt_for_instances` into its own function to make my debugging easier.

Finally, this change turns the node vs service count check into a warning/user prompt. In dev this is something we might be ok with.

Tested this on our dev cluster. Seems reliable now.

### CODE REVIEW

**Comment by mohit** [comment]
Please rebase for master for circle!

**Comment by nathanleiby** [comment]
sadly this one has no tests 😿 (but it does run a few CI steps)

File: src/rollover.py [hunk 1]
```diff
@@ -494,44 +525,42 @@ def run_with_timeout(instance_id, command, timeout):
...
-        OutputS3BucketName = EC2_RUN_OUTPUT_S3_BUCKET,
-        OutputS3KeyPrefix = 'rollover-' + datetime.datetime.now().strftime('%Y%m%d')
+        OutputS3BucketName=EC2_RUN_OUTPUT_S3_BUCKET,
+        OutputS3KeyPrefix='rollover-' + datetime.datetime.now().strftime('%Y%m%d')
     )
 
     # Error sending response
-    if response.get('ResponseMetadata', {}).get('HTTPStatusCode') is not 200:
-        print "ERROR sending command:", response
+    status_code = response.get('ResponseMetadata', {}).get('HTTPStatusCode')
```
Comment (mohit) [inline src/rollover.py]: just curious why _status_code_ was useful to pull out.

File: src/rollover.py [hunk 2]
```diff
@@ -494,44 +525,42 @@ def run_with_timeout(instance_id, command, timeout):
...
+    if status_code is not 200:
+        print "ERROR: sending command:", response
```
Comment (mohit) [inline src/rollover.py]: also print _status_code_?

File: src/rollover.py [hunk 3]
```diff
@@ -494,44 +525,42 @@ def run_with_timeout(instance_id, command, timeout):
...
-        print "Error: could not find command ID in response", response
+        print "ERROR: could not find command ID in response", response
         return False
 
-    invocation_response = client.list_command_invocations(CommandId=command_id, InstanceId=instance_id, Details=True)
-    invocations = invocation_response.get('CommandInvocations')
-    if not invocations:
+    invocation = wait_for_invocation(client, command_id, instance_id, timeout)
+    if not invocation or 'CommandPlugins' not in invocation:
         print "ERROR: failed to run `%s` on instance `%s`" % (command, instance_id)
```
Comment (mohit) [inline src/rollover.py]: any way to get the error here for why this failed?

File: src/rollover.py [hunk 4]
```diff
@@ -494,44 +525,42 @@ def run_with_timeout(instance_id, command, timeout):
...
-
-    # Wait until command reaches final state
-    while result.get('Status') in ['Pending', 'InProgress', 'Cancelling']:
-        time.sleep(2)
-        invocation_response = client.list_command_invocations(CommandId=command_id, InstanceId=instance_id, Details=True)
-        result = invocation_response.get('CommandInvocations', [{}])[0].get('CommandPlugins', [{}])[0]
+    return_code = None
+    for plugin in invocation['CommandPlugins']:
+        if "ResponseCode" in plugin:
+            return_code = plugin['ResponseCode']
```
Comment (mohit) [inline src/rollover.py:563]: Is there a `ResponseBody` that might be useful in case of errors?

**Comment by mohit** [comment]
Added a bunch of questions, hopefully they are not very frustrating. Feel free to just mention if you want to ignore any of them.

File: src/rollover.py (hunk 1 above)
Comment (burnsed) [reply src/rollover.py]: oops, that was just temp debugging. will remove...

File: src/rollover.py (hunk 4 above)
Comment (burnsed) [reply src/rollover.py:563]: Good idea. I've refactored this to return both `return_code`, and `output`

File: src/rollover.py (hunk 3 above)
Comment (burnsed) [reply src/rollover.py]: I clarified the error message here

**Comment by burnsed** [comment]
@mohit great suggestions. They prompted me to refactor `run_with_timeout` a little further.

**Review by mohit — APPROVED** [review]
lgtm

**Pull request merged**